### PR TITLE
copr: add simd for vector distance functions

### PR DESCRIPTION
**Pull request by EricZequan**
### What is changed and how it works?

Issue Number: Ref #17290 

Function Compute is replaced by simd：
- `l2_squared_distance`
- `inner_product`
- `cosine_distance`
- `l2_norm`   

### What's Changed:
After optimization, the performance of low-dimensional vectors does not differ much, and the high-dimensional performance improvement is generally 200% +.The vector distance function calculation before using uncheck_index to get elements. Although in https://github.com/tikv/tikv/pull/17333 and https://github.com/tikv/tikv/pull/17334 we fixed memory alignment and improved bounds checking to optimize performance, we found that it still has room for optimization.

This pr defines the f32 parsing at the time of structured parsing of vectors, avoiding memory alignment problems when dealing with vectors. At the same time, the simd optimization function is used directly in the distance function to perform the computation, which improves the performance.

After optimization, the performance of low-dimensional vectors does not differ much, and **the high-dimensional performance improvement is generally 200%~300% +.**

- l2_squared_distance

| dimension | before       | after        | Speedup (x) |
|-----------|--------------|--------------|-------------|
| 3d        | 2.4573 ns    | 3.3492 ns    | 0.73        |
| 10d       | 4.3755 ns    | 2.6823 ns    | **1.63**    |
| 100d      | 33.421 ns    | 11.157 ns    | **3.00**    |
| 400d      | 206.12 ns    | 59.926 ns    | **3.44**    |
| 784d      | 487.49 ns    | 138.66 ns    | **3.52**    |
| 1000d     | 649.60 ns    | 192.56 ns    | **3.37**    |
| 4000d     | 2,838.3 ns   | 953.03 ns    | **2.98**    |

- l2_distance

| dimension | before       | after        | Speedup (x) |
|-----------|--------------|--------------|-------------|
| 3d        | 2.7286 ns    | 3.3500 ns    | 0.81        |
| 10d       | 4.6754 ns    | 3.0264 ns    | **1.55**    |
| 100d      | 34.306 ns    | 12.446 ns    | **2.76**    |
| 400d      | 206.32 ns    | 62.567 ns    | **3.30**    |
| 784d      | 491.73 ns    | 148.73 ns    | **3.31**    |
| 1000d     | 652.75 ns    | 194.32 ns    | **3.36**    |
| 4000d     | 2,870.2 ns   | 967.67 ns    | **2.97**    |

- inner_product

| dimension | before       | after        | Speedup (x) |
|-----------|--------------|--------------|-------------|
| 3d        | 2.4126 ns    | 3.5134 ns    | 0.69        |
| 10d       | 4.3488 ns    | 2.6561 ns    | **1.64**    |
| 100d      | 32.191 ns    | 10.209 ns    | **3.15**    |
| 400d      | 192.06 ns    | 57.238 ns    | **3.35**    |
| 784d      | 477.57 ns    | 129.00 ns    | **3.70**    |
| 1000d     | 634.12 ns    | 180.39 ns    | **3.52**    |
| 4000d     | 2,805.3 ns   | 978.09 ns    | **2.87**    |

- cosine_distance

| dimension | before       | after        | Speedup (x) |
|-----------|--------------|--------------|-------------|
| 3d        | 3.0262 ns    | 3.2580 ns    | 0.93        |
| 10d       | 6.5512 ns    | 3.4449 ns    | **1.90**    |
| 100d      | 60.630 ns    | 14.494 ns    | **4.18**    |
| 400d      | 278.40 ns    | 64.096 ns    | **4.34**    |
| 784d      | 563.94 ns    | 150.73 ns    | **3.74**    |
| 1000d     | 724.23 ns    | 217.23 ns    | **3.33**    |
| 4000d     | 2,953.6 ns   | 972.29 ns    | **3.04**    |

- l2_norm

| dimension      | before | after | Speedup (x) |
|-----------|---------------|---------------|-------------|
| 3d        | 1.8571 ns     | 3.6121 ns     | 0.51        |
| 10d       | 3.7841 ns     | 3.5989 ns     | **1.05**    |
| 100d      | 33.885 ns     | 12.432 ns     | **2.73**    |
| 400d      | 206.15 ns     | 58.894 ns     | **3.50**    |
| 784d      | 474.14 ns     | 134.61 ns     | **3.52**    |
| 1000d     | 619.50 ns     | 185.18 ns     | **3.35**    |
| 4000d     | 2,680.6 ns    | 977.16 ns     | **2.74**    |                        

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
copr: Improve vector distance performance
```

### CODE REVIEW

File: components/tidb_query_expr/src/impl_vec.rs [hunk 1]
```diff
@@ -69,7 +70,7 @@ mod tests {
             (vec![1.0, 2.0, 3.0], Some(3)),
         ];
         for (arg, expected_output) in cases {
-            let arg = VectorFloat32::from_f32(arg).unwrap();
+            let arg = VectorFloat32::copy_from_f32(arg.as_slice());
```
Comment (breezewish) [inline components/tidb_query_expr/src/impl_vec.rs:73]: A possible simplify:

```suggestion
            let arg = VectorFloat32::copy_from_f32(&arg);
```

File: components/tidb_query_datatype/src/codec/mysql/vector.rs [hunk 2]
```diff
@@ -1,44 +1,35 @@
 // Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
-
-use std::cmp::Ordering;
+#[cfg(target_arch = "aarch64")]
+use core::arch::aarch64::*;
+#[cfg(target_arch = "x86_64")]
+use core::arch::x86_64::*;
```
Comment (breezewish) [inline components/tidb_query_datatype/src/codec/mysql/vector.rs:5]: Why is this needed?

File: components/tidb_query_datatype/src/codec/mysql/vector.rs [hunk 3]
```diff
@@ -227,180 +157,317 @@ impl<'a> VectorFloat32Ref<'a> {
 
     pub fn inner_product(&self, b: VectorFloat32Ref<'a>) -> Result<f64> {
         self.check_dims(b)?;
-        let mut distance: f32 = 0.0;
-        for i in 0..self.len() {
-            distance += unsafe { self.index_unchecked(i) * b.index_unchecked(i) };
-        }
-
-        Ok(distance as f64)
+        let inner_product = SpatialSimilarity::dot(self.data(), b.data())
```
Comment (breezewish) [inline components/tidb_query_datatype/src/codec/mysql/vector.rs:160]: Don't use `expect`, it will panic.

**Comment by EricZequan** [comment]
@breezewish PTAL~

**Review by breezewish — APPROVED** [review]
Looks good!

File: components/tidb_query_datatype/src/codec/mysql/vector.rs [hunk 4]
```diff
@@ -174,233 +109,211 @@ impl<'a> VectorFloat32Ref<'a> {
...
 
-        self.write_u32_le(data.len() as u32)?;
-        self.write_bytes(data.value)?;
+impl<T: BufferWriter> VectorFloat32Encoder for T {}
 
-        Ok(())
-    }
+/// Misaligned, Ref.
+#[derive(Clone, Copy, Debug)]
+pub struct VectorFloat32RefMisaligned<'a> {
```
Comment (gengliqi) [inline components/tidb_query_datatype/src/codec/mysql/vector.rs:208]: The term "unaligned" is usually used instead of "misaligned" to describe data that is not aligned.

**Comment by wuhuizuo** [comment]
It introduced the building failure on linux/arm64 platform.
